apollo-router/src/plugins/authentication.rs

```diff
@@ -189,6 +204,136 @@ async fn get_jwks(url: Url) -> Option<JwkSet> {
     Some(jwks)
 }
 
+#[derive(Debug, Default)]
+struct JWTCriteria {
+    alg: Algorithm,
+    kid: Option<String>,
+}
+
+/// Search the list of JWKS to find a key we can use to decode a JWT.
+///
+/// The search criteria allow us to match a variety of keys depending on which criteria are provided
+/// by the JWT header. The only mandatory parameter is "alg".
+/// Note: "none" is not implemented by jsonwebtoken, so it can't be part of the [`Algorithm`] enum.
+async fn search_jwks(
+    my_jwks: SharedDeduplicate,
+    criteria: &JWTCriteria,
+    jwks_urls: Vec<Url>,
+    context: &Context,
+) -> Result<Option<Jwk>, BoxError> {
+    const HIGHEST_SCORE: usize = 2;
+    let mut candidates = vec![];
+    for jwks_url in jwks_urls {
+        context.enter_active_request().await;
+        // Get the JWKS here
+        let jwks_opt = match my_jwks.get(jwks_url).await {
+            Ok(k) => k,
+            Err(e) => {
+                context.leave_active_request().await;
+                return Err(e.into());
+            }
+        };
+        context.leave_active_request().await;
+        let jwks = jwks_opt.ok_or(Error::JwkSet)?;
+        // Try to figure out if our jwks contains a candidate key (i.e.: a key which matches our
+        // criteria)
+        for mut key in jwks.keys {
+            let mut key_score = 0;
+            // We are only interested in keys which are used for signature verification
+            if let Some(purpose) = &key.common.public_key_use {
+                if purpose != &PublicKeyUse::Signature {
+                    continue;
+                }
+            } else if let Some(purpose) = &key.common.key_operations {
+                if !purpose.contains(&KeyOperations::Verify) {
+                    continue;
+                }
+            } else {
+                continue;
+            };
```

Comment:
nit: maybe that could work better as a filter on `jwks.keys`

Reply:
done: 53b6295c1934a75d51e24fcb8f198576af8f7fb4